Shell/Shell.cpp

```diff
@@ -951,27 +950,93 @@ Vector<String> Shell::expand_globs(const StringView& path, const StringView& bas
     return res;
 }
 
-Vector<String> Shell::expand_parameters(const StringView& param) const
+static void for_each_variable_segment_in_argument(const StringView& arg,
+    Function<void(const StringView&, size_t start_offset, size_t end_offset)> on_variable,
+    Function<void(const StringView&, size_t start_offset, size_t end_offset)> on_non_variable)
 {
-    if (!param.starts_with('$'))
-        return { param };
+    size_t previous_index = 0;
+    Optional<size_t> maybe_index;
+    while (previous_index < arg.length()) {
+        auto maybe_index = arg.substring_view(previous_index, arg.length() - previous_index).find_first_of('$');
+        if (!maybe_index.has_value())
+            break;
 
-    String variable_name = String(param.substring_view(1, param.length() - 1));
-    if (variable_name == "?")
-        return { String::number(last_return_code) };
-    else if (variable_name == "$")
-        return { String::number(getpid()) };
+        auto start_index = maybe_index.value() + previous_index + 1;
+        auto end_index = start_index;
 
-    char* env_value = getenv(variable_name.characters());
-    if (env_value == nullptr)
-        return { "" };
+        if (start_index >= arg.length())
+            break;
 
-    Vector<String> res;
-    String str_env_value = String(env_value);
-    const auto& split_text = str_env_value.split_view(' ');
-    for (auto& part : split_text)
-        res.append(part);
-    return res;
+        if (start_index - 1 > previous_index)
+            on_non_variable(arg.substring_view(previous_index, start_index - previous_index), previous_index, start_index - 1);
+
+        if (arg[end_index] == '$' || arg[end_index] == '?') {
+            // $$ and $?.
+            ++end_index;
+        } else {
+            // Any other variable: $[a-zA-Z0-9_]+.
+            for (; end_index < arg.length(); ++end_index) {
+                if (!is_word_character(arg[end_index]))
+                    break;
+            }
+        }
+        on_variable(arg.substring_view(start_index, end_index - start_index), start_index, end_index);
+        previous_index = end_index;
+    }
+    if (previous_index < arg.length()) {
+        on_non_variable(arg.substring_view(previous_index, arg.length() - previous_index), previous_index, arg.length());
+    }
+}
+
+Vector<String> Shell::expand_parameters(const StringView& param, bool is_inside_string) const
+{
+    Vector<String> expanded;
+    StringBuilder builder;
+
+    for_each_variable_segment_in_argument(
+        param,
+        [&](auto& variable, auto, auto) {
+            if (variable == "?") {
+                builder.append(String::number(last_return_code));
+            } else if (variable == "$") {
+                builder.append(String::number(getpid()));
+            } else {
+                char* env_value = getenv(String { variable }.characters());
```

Comment:
* `$foo` should read from local variables as well as from env
* can we please move this substitution logic for one variable into a separate function? perhaps it could even be a method on something, if we had a better object model than just `Shell` and `Vector<StringView>`.

Reply:
I don't think we have local variables yet, so maybe defer this until we do?
splitting it off is a good idea though, I'll do that.